modyn/supervisor/internal/pipeline_executor/models.py

```diff
@@ -0,0 +1,499 @@
+    # with that logic we avoid having to add disciminator fields to every subclass of StageInfo
+
+    @model_serializer(mode="wrap")
+    def serialize(self, handler: Callable[..., dict[str, Any]]) -> dict[str, Any]:
+        default_serializer = handler(self)
+        return {
+            **default_serializer,
+            "info": self.info.model_dump(by_alias=True) if self.info else None,
+            **({"info_type": self.info.__class__.__name__} if self.info else {}),
+        }
+
+    @model_validator(mode="before")
+    @classmethod
+    def deserializer(cls, data: Any) -> Any:
+        if isinstance(data, dict):
+            if data.get("info_type") and data.get("info"):
+                info_type = cast(type[StageInfo], globals().get(data["info_type"]))
+                data["info"] = info_type.model_validate(data["info"])
+            else:
+                data["info"] = None
+        return data
+
+
+class SupervisorLogs(BaseModel):
+    stage_runs: list[StageLog] = Field(default_factory=list)
+
+    def clear(self) -> None:
+        self.stage_runs.clear()
+
+    def merge(self, logs: list[StageLog]) -> SupervisorLogs:
+        self.stage_runs = self.stage_runs + [run for _l in logs for run in _l.stage_runs]
+        return self
+
+    @property
+    def df(self) -> pd.DataFrame:
+        """Provides a dataframe with log information of all stages which helps to easily plot pipeline run metrics.
+
+        Returns:
+            A DataFrame with the core log information of all stages
+            (not including the additional info property of the StageLogs).
+        """
+
+        return pd.DataFrame(
+            data=[
+                (stage.id, stage.start, stage.end, stage.duration, stage.batch_idx, stage.sample_idx, stage.sample_time)
+                for stage in self.stage_runs
+            ],
+            columns=["id", "start", "end", "duration", "batch_idx", "sample_idx", "sample_time"],
+        )
+
+
+class PipelineLogs(BaseModel):
+    """
+    Wrapped logs for the pipeline executor.
+
+    This file maintains the log directory:
+    logs
+    ├── pipeline_1
+    │   ├── pipeline.log (final logfile being made available at the end of the pipeline execution)
+    │   ├── pipeline.part.log (initial logfile including static fields)
+    │   ├── supervisor.part_0.log (incremental log files)
+    │   ├── supervisor.part_1.log
+    │   └── supervisor.part_2.log
+    └── pipeline_2
+    """
+
+    # static logs
+
+    pipeline_id: int
+    pipeline_stages: dict[str, tuple[int, list[str]]]
+    """List of all pipeline stages, their execution order index, and their parent stages."""
+
+    config: ConfigLogs
+
+    experiment: bool
+    start_replay_at: int | None = Field(None, description="Epoch to start replaying at")
+    stop_replay_at: int | None = Field(None, description="Epoch to stop replaying at")
+
+    # incremental logs
+
+    supervisor: SupervisorLogs = Field(default_factory=SupervisorLogs)
```

Comment:
rename to supervisor_logs? was thinking for a moment that this owns a superviosr object now.

Reply:
Will do! I thought that would be clear from the typing, but fair point.